### PR TITLE
Pulling all Canvas specific grading logic from the services to the views

### PR DESCRIPTION
**Pull request by jon-betts**
Must be merged after https://github.com/hypothesis/lms/pull/1298.

This works by having a callback we pass into the record function which allows us to modify the request.

Some interesting insights as a result of putting the code in the same place so we can see it:

  * We were treating the `ltiLaunchUrl` and `submittedAt` values as independent
    * In truth you get both or you don't get called
  * We were treating the `submittedAt` value like a variable
    * In truth we always set the exact same value

None of this is obvious when the code dealing with this was in two places.

### CODE REVIEW

File: lms/services/lti_outcomes.py [hunk 1]
```diff
@@ -78,25 +72,15 @@ def record_result(  # pylint:disable=no-self-use
...
+            request = pre_record_hook(score=score, request_body=request)
 
-        if lti_launch_url:
-            request["resultRecord"].setdefault("result", {})["resultData"] = {
-                "ltiLaunchUrl": lti_launch_url
-            }
+            if not isinstance(request, dict):
+                raise TypeError(
+                    "The pre-record hook must return the request body as a dict"
+                )
```
Comment (jon-betts) [inline lms/services/lti_outcomes.py]: A little defensive programming is probably in order here. If we end up with lots of hooks we might want a way of calling them that catches argument errors and expects a particular type. Possible raising `BadHookImplementation`, `BadHookReturnType` or something similar.

File: lms/views/api/lti.py [hunk 2]
```diff
@@ -74,38 +74,58 @@ def record_canvas_speedgrader_submission(self):
...
+            "ltiLaunchUrl": self.get_speedgrader_launch_url()
+        }
+
+        #  A `datetime.datetime` that indicates when the submission was
+        # created. This is only used in Canvas and is displayed in the
+        # SpeedGrader as the submission date. If the submission date matches
+        # an existing submission then the existing submission is updated
+        # rather than creating a new submission.
+        request_body["submissionDetails"] = {
+            "submittedAt": datetime.datetime(2001, 1, 1, tzinfo=timezone.utc)
```
Comment (jon-betts) [inline lms/views/api/lti.py]: It's always 2001!

**Comment by jon-betts** [comment]
Hello Sean from the future! This could probably do with having a PR pulled out of it for some of the test changes which weren't directly required but make it much easier to read.

File: lms/services/lti_outcomes.py [hunk 3]
```diff
@@ -45,26 +45,20 @@ def read_result(self, lis_result_sourcedid):  # pylint:disable=no-self-use
             return None
 
     def record_result(  # pylint:disable=no-self-use
-        self, lis_result_sourcedid, score=None, **canvas_extras,
+        self, lis_result_sourcedid, score=None, pre_record_hook=None,
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply lms/services/lti_outcomes.py]: It's a fuzzy area and both terms are used interchangeably (including by me in this PR apparently). Some SO answers:

 * https://stackoverflow.com/questions/11087543/what-is-the-difference-between-hook-and-callback#11087727

To the extent there is a difference I think it's something like:

 * Callbacks are called to inform them of an outcome (so that un-related processing can take place) for example "onSuccess"
 * Hooks are probably a super-set, but include the possibility of modify the process they are called in "preRecord" 
* There are some ancient interpretations of hooks which aren't really applicable in a Python world involving some sort of memory monkey patching to force a method to do something additional

As a result of hooks being a super-set of callbacks, and callbacks being much more common, I think that hooks are often thought of in terms of how they differ from callbacks (intended to modify behavior), but they technically don't imply that.

I'd therefore suggest the following:

 * If we want a term to only include modifying behavior "hook" is correct
 * If we want a term to only include return information "callback" is more specific (but "hook" is still correct)
 * If we want a term to encompass both, only "hook" is correct
 * It's a good idea to add something to a variable to let people know it's a function like `_fn` so it doesn't get mistaken for a regular value:
   * Basically a heads up: "functional programming ahead" to get in the right head space
   * This prevents the "why the hell are there parens on this?" moment
   * ... but we don't want to end-up in C land with `blah_bool`, `blah_str` etc.
 * As "hook" is more generally correct and shorter I'd suggest it's good practice to add it to the end

As for changes in this PR, I should definitely pick a side of the fence and be consistent. So I'll rename references to callbacks to hooks. It's also good to have this discussion because I realised I'd never really

File: lms/services/lti_outcomes.py [hunk 4]
```diff
@@ -45,26 +45,20 @@ def read_result(self, lis_result_sourcedid):  # pylint:disable=no-self-use
...
-            A URL where the student's work on this submission can be viewed.
-            This is only used in Canvas.
-        :param submitted_at:
-        :type datetime.datetime:
-            A `datetime.datetime` that indicates when the submission was
-            created. This is only used in Canvas and is displayed in the
-            SpeedGrader as the submission date. If the submission date matches
-            an existing submission then the existing submission is updated
-            rather than creating a new submission.
+        :param pre_record_hook: The call back
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply lms/services/lti_outcomes.py]: This has been updated with new language (preferring hook)
